crates/task-impls/src/response.rs

```diff
@@ -58,83 +48,63 @@ impl<TYPES: NodeType> NetworkResponseState<TYPES> {
     /// Create the network request state with the info it needs
     pub fn new(
         consensus: LockedConsensusState<TYPES>,
-        receiver: RequestReceiver,
         quorum: Arc<TYPES::Membership>,
         pub_key: TYPES::SignatureKey,
         private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
         id: u64,
     ) -> Self {
         Self {
             consensus,
-            receiver,
             quorum,
             pub_key,
             private_key,
             id,
         }
     }
 
-    /// Run the request response loop until a `HotShotEvent::Shutdown` is received.
-    /// Or the stream is closed.
-    async fn run_loop(mut self, shutdown: EventDependency<Arc<HotShotEvent<TYPES>>>) {
-        let mut shutdown = Box::pin(shutdown.completed().fuse());
+    /// Process request events or loop until a `HotShotEvent::Shutdown` is received.
+    async fn run_response_loop(
+        self,
+        mut receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
+        event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+    ) {
         loop {
-            futures::select! {
-                req = self.receiver.next() => {
-                    match req {
-                        Some((msg, chan)) => self.handle_message(msg, chan).await,
-                        None => return,
-                    }
-                },
-                _ = shutdown => {
-                    return;
-                }
-            }
-        }
-    }
-
-    /// Handle an incoming message.  First validates the sender, then handles the contained request.
-    /// Sends the response via `chan`
-    async fn handle_message(&self, raw_req: Vec<u8>, chan: NetworkMsgResponseChannel<Vec<u8>>) {
-        let req: Message<TYPES> = match bincode::deserialize(&raw_req) {
-            Ok(deserialized) => deserialized,
-            Err(e) => {
-                tracing::error!("Failed to deserialize message! Error: {e}");
-                return;
-            }
-        };
-        let sender = req.sender.clone();
-
-        match req.kind {
-            MessageKind::Data(DataMessage::RequestData(request)) => {
-                if !self.valid_sender(&sender) || !valid_signature::<TYPES>(&request, &sender) {
-                    let serialized_msg = match bincode::serialize(
-                        &self.make_msg(ResponseMessage::Denied),
-                    ) {
-                        Ok(serialized) => serialized,
-                        Err(e) => {
-                            tracing::error!("Failed to serialize outgoing message: this should never happen. Error: {e}");
+            match receiver.recv_direct().await {
```

Comment:
Shouldn't this loop also handle the proposal requests as well now?

Reply:
yes i agree with this. i can move it here